src/apps/lisk/sign_tx.py

```diff
@@ -0,0 +1,126 @@
+from trezor import wire, ui
+from trezor.messages.LiskSignedTx import LiskSignedTx
+from trezor.messages.LiskTransactionType import *
+from trezor.messages import FailureType
+from trezor.utils import HashWriter
+from apps.lisk.layout import *
+from ubinascii import unhexlify, hexlify
+
+
+async def lisk_sign_tx(ctx, msg):
+    from trezor.crypto.hashlib import sha256
+
+    public_key, seckey = await _get_keys(ctx, msg)
+    transaction = update_raw_tx(msg.transaction, public_key)
+
+    # throw ValueError if transaction has not valid structure
+    try:
+        await require_confirm_by_type(ctx, transaction)
+    except AttributeError:
+        raise ValueError(FailureType.DataError, 'The transaction has invalid asset data field')
+
+    await require_confirm_fee(ctx, transaction.amount, transaction.fee)
+
+    sha = HashWriter(sha256)
+    transactionBytes = _get_transaction_bytes(transaction)
+
+    for field in transactionBytes:
+        sha.extend(field)
+
+    digest = sha.get_digest()
+
+    signature = await get_signature(seckey, digest)
+
+    return LiskSignedTx(signature=signature)
+
+async def require_confirm_by_type(ctx, transaction):
+    if transaction.type is Transfer:
+        return await require_confirm_tx(ctx, transaction.recipient_id, transaction.amount)
+    if transaction.type is RegisterDelegate:
+        return await require_confirm_delegate_registration(ctx, transaction.asset.delegate.username)
+    if transaction.type is CastVotes:
+        return await require_confirm_vote_tx(ctx, transaction.asset.votes)
+    if transaction.type is RegisterSecondPassphrase:
+        return await require_confirm_public_key(ctx, transaction.asset.signature.public_key)
+    if transaction.type is RegisterMultisignatureAccount:
+        return await require_confirm_multisig(ctx, transaction.asset.multisignature)
+
+async def get_signature(seckey, digest):
+    from trezor.crypto.curve import ed25519
+    signature = ed25519.sign(seckey, digest)
+
+    return signature
+
+def _get_transaction_bytes(msg):
+    from ustruct import pack
+
+    # Required transaction parameters
+    t_type = pack('<b', msg.type)
+    t_timestamp = pack('<i', msg.timestamp)
+    t_amount = pack('<Q', msg.amount)
+    t_pubkey = msg.sender_public_key
+
+    if msg.requester_public_key is None:
+        t_requester_public_key = b''
+    else:
+        t_requester_public_key = msg.requester_public_key
+
+    if msg.recipient_id is None:
+        t_recipient_id = pack('>Q', 0)
+    else:
+        t_recipient_id = pack('>Q', int(msg.recipient_id[:-1]))
+
+    if msg.signature is None:
+        t_signature = b''
+    else:
+        t_signature = msg.signature
+
+    t_asset = _get_asset_data_byttes(msg)
+
+    return [t_type, t_timestamp, t_pubkey, t_requester_public_key, t_recipient_id, t_amount, t_asset, t_signature]
+
+def _get_asset_data_byttes(msg):
+    from ustruct import pack
+    data = b''
+
+    if msg.type is Transfer and getattr(msg.asset, "data"):
```

Comment:
the `getattr` here is pointless; it will raise the same `AttributeError` that the read of `msg.asset.data` would on the following line.
if you want to check for presence of `data`, you should make that clear by saying `msg.asset.data is not None`

btw, this is not covered by the `require_confirm_by_type` call where you're catching `AttributeError`s like this.
maybe the try/catch above should extend to covert the `_get_transaction_bytes` call too?